### PR TITLE
RavenDB-15330 - Put raft commands with `DontCareId` in the history log as well

### PR DESCRIPTION
**Pull request by shaharhikri**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-15330/Put-raft-commands-with-DontCareId-in-the-history-log-as-well

### Additional description

Put raft commands with `DontCareId` in the history log as well

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Rachis/RachisLogHistory.cs [hunk 1]
```diff
@@ -134,9 +134,6 @@ public void InsertHistoryLog(ClusterOperationContext context, long index, long t
             if (guid == null) // shouldn't happened in new cluster version!
                 return;
 
-            if (guid == RaftIdGenerator.DontCareId)
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/RachisLogHistory.cs]: You can't do that.

A command is uniquely defined by this guid, so All `DontCareId` commands will have the same string (the empty one) and will be considered as the same command... which will break everything :)

**Review by karmeli87 — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Rachis/RachisLogHistory.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Server/Rachis/RachisLogHistory.cs]: Although I see all tests has passed.. this is because you only removed it here in the `InsertHistoryLog` so this probably will not be updated to the status of `Committed`.. 

Maybe we can generate a stable id here... let's say something like `DontCare/{index}`?

Comment (shaharhikri) [reply src/Raven.Server/Rachis/RachisLogHistory.cs]: Done

File: src/Raven.Server/Rachis/RachisLogHistory.cs [hunk 2]
```diff
@@ -135,7 +135,9 @@ public void InsertHistoryLog(ClusterOperationContext context, long index, long t
                 return;
 
             if (guid == RaftIdGenerator.DontCareId)
+            {
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/RachisLogHistory.cs]: remove unnecessary changes

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 3]
```diff
@@ -1489,6 +1491,17 @@ public long AppendToLog(CommandBase cmd, long term)
...
             }
             lastIndex += 1;
+            var guid = LogHistory.GetGuidFromCommand(cmd);
+            if (guid == RaftIdGenerator.DontCareId)
+            {
+                var newGuid = $"DontCare/{lastIndex}";
+                cmd.Modifications = new DynamicJsonValue { [nameof(CommandBase.UniqueRequestId)] = newGuid };
+                using (cmd)
+                {
+                    cmd = context.ReadObject(cmd, newGuid);
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/RachisConsensus.cs:1501]: please fix this warning

File: test/SlowTests/Issues/RavenDB-15330.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+{
+    public RavenDB_15330(ITestOutputHelper output) : base(output)
+    {
+    }
+
+    [Fact]
+    public async Task Command_With_DontCareId_Should_Be_Committed()
+    {
+        DebuggerAttachedTimeout.DisableLongTimespan = true;
+        var server = GetNewServer();
```
Comment (karmeli87) [inline test/SlowTests/Issues/RavenDB-15330.cs:40]: let's use `DoNotReuseServer` method instead

File: test/SlowTests/Issues/RavenDB-15330.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        using (server.ServerStore.Engine.ContextPool.AllocateOperationContext(out ClusterOperationContext context))
+        using (context.OpenReadTransaction())
+        {
+            var historyLog = server.ServerStore.Engine.LogHistory.GetHistoryLogs(context);
+            bool contain = false;
+            foreach (var djv in historyLog)
+            {
+                if (djv!=null && 
+                    djv["Type"] != null && djv["Type"].ToString() == nameof(PutRollingIndexCommand) && 
+                    djv["State"]!=null && djv["State"].ToString() == "Committed")
```
Comment (karmeli87) [inline test/SlowTests/Issues/RavenDB-15330.cs]: use `nameof`

File: test/SlowTests/Issues/RavenDB-15330.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        var result = await server.ServerStore.SendToLeaderAsync(new PutRollingIndexCommand(store.Database, nameof(SomeRollingIndex),
+            DateTime.UtcNow, RaftIdGenerator.DontCareId));
+
+        using (server.ServerStore.Engine.ContextPool.AllocateOperationContext(out ClusterOperationContext context))
+        using (context.OpenReadTransaction())
+        {
+            var historyLog = server.ServerStore.Engine.LogHistory.GetHistoryLogs(context);
+            bool contain = false;
+            foreach (var djv in historyLog)
+            {
```
Comment (karmeli87) [inline test/SlowTests/Issues/RavenDB-15330.cs:62]: after this change I expect to see all entries in the log without any gaps.
Let's check this as well.

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 7]
```diff
@@ -1489,6 +1491,17 @@ public long AppendToLog(CommandBase cmd, long term)
                 GetLastTruncated(context, out lastIndex, out long _);
             }
             lastIndex += 1;
+            var guid = LogHistory.GetGuidFromCommand(cmd);
+            if (guid == RaftIdGenerator.DontCareId)
+            {
+                var newGuid = $"DontCare/{lastIndex}";
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/RachisConsensus.cs]: after second thought, in order this to be unique the index isn't enough, it might happened that we will have same index on different terms, so let's add the term here as well.

**Review by karmeli87 — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Rachis/RachisLogHistory.cs (hunk 2 above)
Comment (shaharhikri) [reply src/Raven.Server/Rachis/RachisLogHistory.cs]: Done

File: src/Raven.Server/Rachis/RachisConsensus.cs (hunk 3 above)
Comment (shaharhikri) [reply src/Raven.Server/Rachis/RachisConsensus.cs:1501]: Done

File: test/SlowTests/Issues/RavenDB-15330.cs (hunk 4 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-15330.cs:40]: Done

File: test/SlowTests/Issues/RavenDB-15330.cs (hunk 5 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-15330.cs]: Done

File: test/SlowTests/Issues/RavenDB-15330.cs (hunk 6 above)
Comment (shaharhikri) [reply test/SlowTests/Issues/RavenDB-15330.cs:63]: Done

File: src/Raven.Server/Rachis/RachisConsensus.cs (hunk 7 above)
Comment (shaharhikri) [reply src/Raven.Server/Rachis/RachisConsensus.cs]: Done

File: test/SlowTests/Issues/RavenDB-15330.cs (hunk 4 above)
Comment (karmeli87) [reply test/SlowTests/Issues/RavenDB-15330.cs:40]: if you have `DoNotReuseServer`, do you still need this server?

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 8]
```diff
@@ -1489,6 +1491,19 @@ public long AppendToLog(CommandBase cmd, long term)
...
+            if (guid == RaftIdGenerator.DontCareId)
+            {
+                var newGuid = $"DontCare/{term}-{lastIndex}";
+                cmd.Modifications = new DynamicJsonValue { [nameof(CommandBase.UniqueRequestId)] = newGuid };
+                BlittableJsonReaderObject newCmd;
+                using (cmd)
+                {
+                    newCmd = context.ReadObject(cmd, newGuid);
+                }
+                cmd = newCmd;
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/RachisConsensus.cs]: a bit nicer way :)
```
using (var old = cmd)
{
    cmd = context.ReadObject(cmd, newGuid);
}
```

**Review by karmeli87 — APPROVED** [review]

File: src/Raven.Server/Rachis/RachisConsensus.cs (hunk 8 above)
Comment (shaharhikri) [reply src/Raven.Server/Rachis/RachisConsensus.cs]: Done

**Comment by shaharhikri** [comment]
https://github.com/ravendb/ravendb/pull/16013#discussion_r1142399212
I removed `DoNotReuseServer`

**Review by karmeli87 — APPROVED** [review]
